src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicenseExpressionParser.cs

```diff
@@ -0,0 +1,201 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+
+namespace NuGet.Packaging.Licenses
+{
+    public static class NuGetLicenseExpressionParser
+    {
+        /// <summary>
+        /// Parses a License Expression if valid.
+        /// The expression would be parsed correct, even if non-standard exceptions are encountered. The non-standard Licenses/Exceptions have metadata on them with which the caller can make decisions.
+        /// Based on the Shunting Yard algorithm. <see href="https://en.wikipedia.org/wiki/Shunting-yard_algorithm"/>
+        /// This method first creates an postfix expression by separating the operators and operands.
+        /// Later the postfix expression is evaluated into an object model that represents the expression. Note that brackets are dropped in this conversion and this is not round-trippable.
+        /// The token precedence helps make sure that the expression is a valid infix one. 
+        /// </summary>
+        /// <param name="expression"></param>
+        /// <returns>NuGetLicenseExpression</returns>
+        /// <exception cref="ArgumentException">If the expression has invalid characters</exception>
+        /// <exception cref="ArgumentException">If the expression is empty or null.</exception>
+        /// <exception cref="ArgumentException">If the expression itself is invalid. Example: MIT OR OR Apache-2.0, or the MIT or Apache-2.0, because the expressions are case sensitive.</exception>
+        /// <exception cref="ArgumentException">If the expression's brackets are mismatched.</exception>
+        /// <exception cref="ArgumentException">If the licenseIdentifier is deprecated.</exception>
+        /// <exception cref="ArgumentException">If the exception identifier is deprecated.</exception>
+        public static NuGetLicenseExpression Parse(string expression)
```

Comment:
not sure why there are separate impl for tokens and then parsing the whole expression which also jut make sure if that's a standard expression or complex one and with valid license identifiers.

I believe the earlier comment and logic to parse this expression should work.

Reply:
Refer to the earlier comment. 

Having a tokenizer and lexer/parser is a common pattern for lexical analyzers.